lua/mason-null-ls/init.lua

```diff
@@ -1,140 +1,135 @@
-		table[item] = true
+	-- NOTE: this is left here for future porting in case needed
+	-- local ok, err = pcall(function()
+	--     require "mason-lspconfig.lspconfig_hook"()
+	--     require "mason-lspconfig.server_config_extensions"()
+	-- end)
+	-- if not ok then
+	--     log.error("Failed to set up lspconfig integration.", err)
+	-- end
+
+	if #settings.current.ensure_installed > 0 then
+		require('mason-null-ls.ensure_installed')()
 	end
 
-	return vim.tbl_keys(table)
+	require('mason-null-ls.api.command')
 end
 
-local function lookup(t)
-	local tools = {}
-	for _, source in ipairs(t) do
-		local wantedTools = mappings[source] or {}
-		for _, tool in pairs(wantedTools) do
-			tools[tool] = true
+---See `:h mason-null-ls.setup_handlers()`
+---@param handlers table<string, fun(source_name: string)>
+function M.setup_handlers(handlers)
+	local Optional = require('mason-core.optional')
+	local source_mapping = require('mason-null-ls.mappings.source')
+	local registry = require('mason-registry')
+	local notify = require('mason-core.notify')
+
+	local default_handler = Optional.of_nilable(handlers[1])
+
+	_.each(function(handler)
+		if type(handler) == 'string' and not source_mapping.null_ls_to_package[handler] then
+			notify(
+				('mason-null-ls.setup_handlers: Received handler for unknown null-ls source name: %s.'):format(handler),
+				vim.log.levels.WARN
+			)
 		end
+	end, _.keys(handlers))
+
+	---@param pkg_name string
+	local function get_source_name(pkg_name)
+		return Optional.of_nilable(source_mapping.package_to_null_ls[pkg_name])
 	end
-	return vim.tbl_keys(tools)
-end
 
-local setup = function(settings)
-	SETTINGS = vim.tbl_deep_extend('force', SETTINGS, settings)
-	vim.validate({
-		ensure_installed = { SETTINGS.ensure_installed, 'table', true },
-		null_ls_sources = { SETTINGS.null_ls_sources, 'table', true },
-		auto_update = { SETTINGS.auto_update, 'boolean', true },
-		automatic_installation = { SETTINGS.automatic_installation, 'boolean', true },
-	})
-end
+	local function call_handler(source_name)
+		log.fmt_trace('Checking handler for %s', source_name)
+		Optional.of_nilable(handlers[source_name]):or_(_.always(default_handler)):if_present(function(handler)
+			log.fmt_trace('Calling handler for %s', source_name)
+			local ok, err = pcall(handler, source_name)
+			if not ok then
+				vim.notify(err, vim.log.levels.ERROR)
+			end
+		end)
+	end
 
-local show = function(msg)
-	vim.schedule_wrap(print(string.format('[mason-null-ls] %s', msg)))
+	local installed_sources = _.filter_map(get_source_name, registry.get_installed_package_names())
+	_.each(call_handler, installed_sources)
+	registry:on(
+		'package:install:success',
+		vim.schedule_wrap(function(pkg)
+			get_source_name(pkg.name):if_present(call_handler)
+		end)
+	)
 end
 
-local show_error = function(msg)
-	vim.schedule_wrap(vim.api.nvim_err_writeln(string.format('[mason-null-ls] %s', msg)))
-end
+---@return string[]
+function M.get_installed_sources()
```

Comment:
that makes sense

Reply:
I might be misunderstanding the comment (do you mean that it may return sources that are installed outside of mason?). By just looking at the code quickly, it should only ever return sources that are installed via mason.

Reply:
no, I meant users would probably want to know tools installed by this `mason-null-ls` plugin. not all tools that can be used with a null-ls source, which is what is currently happening

example, I do 

```
require("mason-null-ls").setup({
    ensure_installed = { "stylua", "jq" }
})
```

this will install `stylua` and `jq` from mason. I also do `:Mason` and install `editorconfig-checker`. calling this function will return all 3 things while I only installed 2 using this tool.

Reply:
`Disclaimer: This function will return tools that may not have been installed by this plugin. It will occur when a mason tool matches a source available in `null-ls`

Reply:
To be fair that's a feature of mason extensions, for instance `mason-lspconfig` does automatically handle lsp servers installed by mason.

> ... that may not have been installed by this plugin

After all, this plugin isn't installing anything itself, it translates names and asks mason to install stuff.